distro.py

```diff
@@ -1074,7 +1074,72 @@ def _parse_distro_release_content(line):
         return distro_info
 
 
-_distro = LinuxDistribution()
+def _normalize_id(distro_id, table):
+    """ Helper function which normalizes a distro id value. """
+    distro_id = distro_id.lower().replace(' ', '_')
+    return table.get(distro_id, distro_id)
+
+
+class LinuxMintDistribution(LinuxDistribution):
+    """ LinuxMint will have the same /etc/os-release file
+    as it's upstream Ubuntu so we should not use any info
+    from that file except `ID_LIKE`. """
+    def os_release_attr(self, attribute):
+        if attribute == 'id_like':
+            v = super(LinuxMintDistribution, self).os_release_attr(attribute)
+            return v
+        return ''
+
+
+class CloudLinuxDistribution(LinuxDistribution):
+    """ CloudLinux <7 doesn't provide an lsb-release or os-release file,
+    only a redhat-release file so it's id is incorrectly detected as `rhel`.
+    """
+    def id(self):
+        return 'cloudlinux'
+
+
+class DebianDistribution(LinuxDistribution):
+    """ Starting with Debian 9 (stretch) the information inside of
+    /etc/os-release does not follow the same standards for codenames
+    that all other distros follow thus custom detection of Debians
+    codename is required. """
+    def _parse_os_release_content(self, lines):
+        props = (super(DebianDistribution, self).
+                 _parse_os_release_content(lines))
+        if 'codename' not in props and props.get('pretty_name', ''):
+            pretty_name = props['pretty_name']
+
+            # Remove the NAME from the PRETTY_NAME entry to prevent getting
+            # a codename of 'GNU/Linux'.
+            pretty_name = pretty_name.replace(props.get('name', ''), '')
+
+            match = re.search(r'\s([^\s\d]+)$', pretty_name)
+            if match:
+                props['codename'] = match.group(1)
+        return props
+
+
+def get_implementation(*args):
```

Comment:
It's public-ish. It's probably not going to be used by a single non-power user because most users will just use `distro.id()`, `distro.name()`, etc. The word implementation was chosen because @nir0s suggested it in a prior thread. I thought that naming it `get_distribution` would clash too closely to the other function that I know will be defined `distribution()` when we add in support for Windows and Mac OS.